radix-engine/src/system/transaction/multithread_intent_processor.rs

```diff
@@ -0,0 +1,189 @@
+
+            let mut system_service = SystemService::new(api);
+            let simulate_every_proof_under_resources = intent
+                .auth_zone_init
+                .simulate_every_proof_under_resources
+                .clone();
+            let initial_non_fungible_id_proofs =
+                intent.auth_zone_init.initial_non_fungible_id_proofs.clone();
+            let auth_zone = AuthModule::create_auth_zone(
+                &mut system_service,
+                None,
+                simulate_every_proof_under_resources,
+                initial_non_fungible_id_proofs,
+            )?;
+
+            api.kernel_set_call_frame_data(Actor::Function(FunctionActor {
+                blueprint_id: BlueprintId::new(
+                    &TRANSACTION_PROCESSOR_PACKAGE,
+                    TRANSACTION_PROCESSOR_BLUEPRINT,
+                ),
+                ident: TRANSACTION_PROCESSOR_RUN_IDENT.to_string(),
+                auth_zone,
+            }))?;
+
+            let mut system_service = SystemService::new(api);
+            let txn_processor = IntentProcessor::<InstructionV2>::init(
+                intent.encoded_instructions.clone(),
+                global_address_reservations.clone(),
+                intent.blobs.clone(),
+                MAX_TOTAL_BLOB_SIZE_PER_INVOCATION,
+                &mut system_service,
+            )?;
+
+            txn_processors.push((
+                txn_processor,
+                intent
+                    .children_subintent_indices
+                    .iter()
+                    .map(|index| {
+                        let thread_index = index.0 + 1;
+                        thread_index
+                    })
+                    .collect::<Vec<_>>(),
+            ));
+        }
+        Ok(Self {
+            threads: txn_processors,
+        })
+    }
+
+    pub fn execute<Y: SystemBasedKernelApi>(&mut self, api: &mut Y) -> Result<(), RuntimeError> {
+        let mut cur_thread = 0;
+        let mut parent_stack = vec![];
+        let mut passed_value = None;
+
+        loop {
+            api.kernel_switch_stack(cur_thread)?;
+            let (txn_thread, children_mapping) = self.threads.get_mut(cur_thread).unwrap();
+
+            let mut system_service = SystemService::new(api);
+            let switch_to = match txn_thread.resume(passed_value.take(), &mut system_service)? {
+                ResumeResult::YieldToChild(child, value) => {
+                    let child = *children_mapping.get(child).unwrap();
+                    parent_stack.push(cur_thread);
+                    Some((child, value, false))
+                }
+                ResumeResult::YieldToParent(value) => {
+                    let parent = parent_stack.pop().unwrap();
+                    Some((parent, value, false))
+                }
+                ResumeResult::ChildIntentDone(value) => {
+                    let parent = parent_stack.pop().unwrap();
+                    Some((parent, value, true))
```

Comment:
I think it would be great if we stored a state `ThreadState::Active / ThreadState::Complete` against each thread, and give a `RuntimeError` if:
* Child intents are resumed when complete
* Any thread except thread 0 returns `ResumeResult::RootIntentDone`
* Thread 0 returns `ResumeResult::ChildIntentDone`
* At the end, if any threads are not `ThreadState::Complete`

These would all form extra sanity checks - and would be useful to catch issues in tests if we use a `TestTransaction` which purposefully doesn't undergo validation.

Reply:
I'll add this as a TODO comment as it's not strictly necessary and probably better implemented when we write such tests.